dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+    int* __restrict__ idx_workspace, typename SampleDescT::In* __restrict__ input_workspace,
+    Cb&& cb) {
+  if (sample_desc.shape.lanes_axis == 2) {
+    cb(create_shm_conv<2>(sample_desc, in_loader, block_setup, input_workspace, idx_workspace));
+  } else {
+    assert(sample_desc.shape.lanes_axis == 1);
+    cb(create_shm_conv<1>(sample_desc, in_loader, block_setup, input_workspace, idx_workspace));
+  }
+}
+
+template <int lanes_axis, typename SampleDescT, typename Inloader, typename BlockSetupT>
+DALI_DEVICE DALI_FORCEINLINE DirectInputConv<lanes_axis, SampleDescT, Inloader, BlockSetupT>
+create_direct_conv(const SampleDescT& sample_desc, const Inloader& in_loader,
+                   const BlockSetupT& block_setup) {
+  return {sample_desc, in_loader, block_setup};
+}
+
+template <typename SampleDescT, typename InLoaderT, typename BlockSetupT, typename Cb>
+DALI_DEVICE DALI_FORCEINLINE std::enable_if_t<SampleDescT::axes == 2, void> with_direct_conv(
+    const SampleDescT& sample_desc, const InLoaderT& in_loader, const BlockSetupT& block_setup,
+    Cb&& cb) {
+  cb(create_direct_conv<1>(sample_desc, in_loader, block_setup));
+}
+
+template <typename SampleDescT, typename InLoaderT, typename BlockSetupT, typename Cb>
+DALI_DEVICE DALI_FORCEINLINE std::enable_if_t<SampleDescT::axes == 3, void> with_direct_conv(
+    const SampleDescT& sample_desc, const InLoaderT& in_loader, const BlockSetupT& block_setup,
+    Cb&& cb) {
+  if (sample_desc.shape.lanes_axis == 2) {
+    cb(create_direct_conv<2>(sample_desc, in_loader, block_setup));
+  } else {
+    assert(sample_desc.shape.lanes_axis == 1);
+    cb(create_direct_conv<1>(sample_desc, in_loader, block_setup));
+  }
+}
+
+/*
+Given a HWC image and RS filter, all the necessary products for computing the convolution
+explicitly can be seen as multiplying a matrix of shape HWC x RS with a vector of size RS. Now,
+assuming a standard contiguious memory layout of the HWC image, if you look at any given column of
+the HWC x RS matrix, consecutive rows map to consecutive memory addresses (with the exception of
+positions when we cross H, W extents and border remapping takes place). This implementation melds
+the W and C extents to utilize this property. Additionally, the implementation that uses shared
+memory loads the inputs in blocks with extents corresponding to (D, )H, W * C extents to account
+for the fact that spatailly close products will reuse some of the inputs.
+*/
+template <typename SampleDescT, typename BlockSetupFactory, typename InLoaderFactory,
+          typename GridSetupT>
+__global__ void filter(const SampleDescT* __restrict__ descs, BlockSetupFactory block_setup_factory,
+                       InLoaderFactory in_loader_factory, GridSetupT grid_setup) {
+  extern __shared__ char shm[];
+  int sample_idx = grid_setup.sample_idx();
+  for (int sample_idx = grid_setup.sample_idx(); sample_idx < grid_setup.num_samples();
+       sample_idx += grid_setup.sample_dim()) {
```

Comment:
sample_dim looks a bit weird in this context, I would expect it is somehow related to the dimension/ndim.

Reply:
Renamed it to max_num_samples.